### DISCUSSION TITLE
package.json prevents vercel dev from working despite vercel.json

### DISCUSSION THREAD

**Discussion by jambonrose** [vercel/community#3378] [Help] [answered]
### Summary

I'm migrating a static website (HTML & CSS, no JS or framework) to Vercel, with the intent of eventually replacing the hand-hacked code with a framework.

I've successfully deployed the site via the Github integration, and I am also able to run `vercel dev` to see changes locally. To achieve this, I have created the following `vercel.json`.

```json
{
	"framework": null,
	"installCommand": null,
	"buildCommand": null,
	"outputDirectory": "public"
}
```

I now want to install Playwright so I can write tests to ensure that the framework I will port to properly mimics the existing site. I use `pnpm init` to create a `package.json` file.

However, the moment I create the `package.json` file, running `vercel dev` ceases to work. Instead, I am shown:

```
Vercel CLI 31.1.0
Error: Your `package.json` file is missing a `build` property inside the `scripts` property.
Learn More: https://vercel.link/missing-build-script
```

I note that the [documentation](https://vercel.com/docs/errors/error-list#missing-build-script) seems to imply that a "build" script is only required if no `vercel.json` file exists, but I definitely have one.

I note that when I push the branch to Github with the `package.json` file Vercel still deploys the site without problem. I believe the problem is limited to `vercel dev`.

How may I use `vercel dev` in tandem with Playwright when I have no framework or build step? Adding `"build": null,` or `"build": "",` to `package.json` has not worked for me.

### Steps to Reproduce

```console
$ mkdir mve && cd mve
$ mkdir public
$ touch public/index.html
$ echo '<body><p>hello world</p></body>' > public/index.html
$ touch vercel.json
$ cat << EOF > vercel.json
{
        "framework": null,
        "installCommand": null,
        "buildCommand": null,
        "outputDirectory": "public"
}
EOF
```

For reasons I don't understand, the first time I run `vercel dev` the setup fails, but a subsequent call works without changes. I note that the build information listed on the Vercel dashboard shows no errors.

```console
$ vercel dev
Vercel CLI 31.1.0
? Set up and develop “~/Development/mve”? [Y/n] y
? Which scope should contain your project? jambonrose
? Link to existing project? [y/N] n
? What’s your project’s name? mve
? In which directory is your code located? ./public
🔗  Linked to jambonrose/mve (created .vercel and added it to .gitignore)
🔍  Inspect: https://vercel.com/jambonrose/mve/EhW1MNaYAWifVEqANRpQ9dKvYKBJ [3s]
Error: Failed to detect project settings. Please try again.

$ vercel dev
Vercel CLI 31.1.0
> Ready! Available at http://localhost:3000
```

Everything works! Now let's break it.

```console
$ pnpm init
$ vercel dev
Vercel CLI 31.1.0
Error: Your `package.json` file is missing a `build` property inside the `scripts` property.
Learn More: https://vercel.link/missing-build-script
```

**Comment by amyegan** [comment, marked as answer]
Hey @jambonrose. Thanks sharing this scenario. I got it to work in my test app by setting `buildCommand` to `""` instead of `null` to explicitly override the default command. Without that, it was falling back to the default which tries to run whatever `build` command is in  the package.json file.

I'll let the CLI team know about this to see what we can do to make the `vercel.json` overrides easier to use with `vercel dev`

**Comment by jambonrose** [comment]
Thanks @amyegan .

I confirm that modifying `vercel.json` to the following fixes the issue when a `package.json` exists.

```json
{
	"framework": null,
	"installCommand": null,
	"buildCommand": "",
	"outputDirectory": "public"
}
```

I note that there's some inconsistency in how `""` and `null` are handled. When I change `"framework"` to `""`, I get the following error:

```
Error: Invalid request: `projectSettings.framework` should be equal to one of the allowed values "blitzjs, nextjs, gatsby, remix, astro, hexo, eleventy, docusaurus-2, docusaurus, preact, solidstart, dojo, ember, vue, scully, ionic-angular, angular, polymer, svelte, sveltekit, sveltekit-1, ionic-react, create-react-app, gridsome, umijs, sapper, saber, stencil, nuxtjs, redwoodjs, hugo, jekyll, brunch, middleman, zola, hydrogen, vite, vitepress, vuepress, parcel, sanity, storybook, ".
```

Thanks for your help!
